src/ray/raylet_client/raylet_client.h

```diff
@@ -231,6 +230,52 @@ class RayletConnection {
   std::mutex write_mutex_;
 };
 
+// Batches PinObjectIDsRequest so there would be only one outstanding
+// request per Raylet. This reduces the memory and CPU overhead when a
+// large number of objects need to be pinned.
+class PinBatcher {
+ public:
+  PinBatcher(std::shared_ptr<rpc::NodeManagerWorkerClient> grpc_client);
+
+  // Adds objects to be pinned at the address.
+  void Add(const rpc::Address &address,
+           std::vector<ObjectID> object_ids,
+           rpc::ClientCallback<rpc::PinObjectIDsReply> callback);
+
+  // Total number of objects waiting to be pinned.
+  int64_t TotalPending() const;
+
+ private:
+  // Request from a single Add() call.
+  struct Request {
+    Request(std::vector<ObjectID> oid, rpc::ClientCallback<rpc::PinObjectIDsReply> cb)
+        : object_ids(std::move(oid)), callback(std::move(cb)) {}
+
+    std::vector<ObjectID> object_ids;
+    rpc::ClientCallback<rpc::PinObjectIDsReply> callback;
+  };
+
+  struct RayletDestination {
+    RayletDestination(PinBatcher *batcher, const rpc::Address &address)
+        : pin_batcher_(batcher), raylet_address_(address) {}
+
+    // Tries sending out a request, if there are buffered messages but no
+    // request is inflight.
+    bool Flush() ABSL_EXCLUSIVE_LOCKS_REQUIRED(pin_batcher_->mu_);
+
+    PinBatcher *const pin_batcher_;
```

Comment:
Let's try to avoid using raw pointers. I think we can potentially make this code structure nicer if we avoid giving the internal RayletDestination a pointer to pin_batcher_. Instead of having the `PinBatcher::RayletDestination::Flush()` method, we can do `PinBatcher::Flush(std::string raylet_id)`.

Reply:
Good suggestion!

Reply:
Hmm is this part of the code updated?

Reply:
Just pushed.